subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/ArtifactViewAttributesIntegrationTest.groovy

```diff
@@ -0,0 +1,382 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.attributes
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Ignore
+
+/**
+ * Tests for [org.gradle.api.artifacts.ArtifactView ArtifactView] that ensure it uses the "live" attributes
+ * of the configuration that created it to select files and artifacts.
+ *
+ * These tests assume a situation where a consumer should resolve the secondary "classes" variant of a producer
+ * and obtain the "main" directory as the only file.  As the necessary attributes are not present on the
+ * consumer's compileClasspath configuration until the beforeLocking callback method is called, the previous
+ * incorrect behavior of the ArtifactView was to select the standard jar file variant instead of the "classes"
+ * variant.
+ */
+class ArtifactViewAttributesIntegrationTest extends AbstractIntegrationSpec {
+    private static declareIncomingFiles = "def incomingFiles = configurations.compileClasspath.incoming.files"
+    private static declareIncomingArtifacts = "def incomingArtifacts = configurations.compileClasspath.incoming.artifacts"
+    private static declareArtifactViewFiles = "def artifactViewFiles = configurations.compileClasspath.incoming.artifactView { }.files"
+    private static declareArtifactViewArtifacts = "def artifactViewArtifacts = configurations.compileClasspath.incoming.artifactView { }.artifacts"
+
+    private static iterateIncomingFiles = """logger.warn 'Incoming Files:'
+incomingFiles.each {
+    logger.warn 'Name: ' + it.name
+}
+logger.warn ''
+"""
+    private static iterateIncomingArtifacts = """logger.warn 'Incoming Artifacts:'
+incomingArtifacts.each {
+    logger.warn 'Name: ' + it.id.name + ', File: ' + it.id.file.name
+}
+logger.warn ''
+"""
+    private static iterateArtifactViewFiles = """logger.warn 'Artifact View Files:'
+artifactViewFiles.each {
+    logger.warn 'Name: ' + it.name
+}
+logger.warn ''
+"""
+    private static iterateArtifactViewArtifacts = """logger.warn 'Artifact View Artifacts:'
+artifactViewArtifacts.each {
+    logger.warn 'Name: ' + it.id.name + ', File: ' + it.id.file.name
+}
+logger.warn ''
+"""
+
+    private static filesComparison = "assert incomingFiles*.name == artifactViewFiles*.name"
+    private static artifactComparison = "assert incomingArtifacts*.id.file.name == artifactViewArtifacts*.id.file.name"
+
+    def setup() {
+        settingsFile << """
+            rootProject.name = "consumer"
+            include "producer"
+        """
+
+        file("producer/build.gradle").text = """
+            plugins {
+                id 'java-library'
```

Comment:
I think this test deserves a simple test-case which does not rely on java library. Something like (pseudocode/untested):

```
        file("producer/build.gradle") << """
            def attr = Attribute.of("test-attr", String)

            task task1 {
                outputs.file file('file1')
            }

            task task2 {
                outputs.file file('file2')
            }

            configurations {
                conf1 {
                    attributes {
                        attribute(attr, 'foo')
                    }
                    outgoing {
                        artifact tasks.task1
                    }
                }
                conf2 {
                    attributes {
                        attributes(attr, 'bar')
                    }
                    outgoing {
                        artifact tasks.task2
                    }
                }
            }
        """

        buildFile << """
            def attr = Attribute.of("test-attr", String)
            
            configurations {
                conf {
                    attributes {
                        attribute(attr, 'foo')
                    }
                }
            }

            dependencies {
                conf(":producer") 
            }

            task resolve {
                def before = configurations.conf.incoming.artifacts
                configurations.conf.attributes.attribute(attr, 'bar')
                def after = configurations.conf.incoming.artifacts
                doLast {
                    assert before.files == after.files
                }
            }
        """


```

Reply:
That's a good idea.